presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -825,6 +832,49 @@ protected Scope visitTableSubquery(TableSubquery node, Optional<Scope> scope)
         @Override
         protected Scope visitQuerySpecification(QuerySpecification node, Optional<Scope> scope)
         {
+            if (SystemSessionProperties.isLegacyOrderByEnabled(session)) {
+                return legacyVisitQuerySpecification(node, scope);
+            }
+
+            // TODO: extract candidate names from SELECT, WHERE, HAVING, GROUP BY and ORDER BY expressions
+            // to pass down to analyzeFrom
+
+            Scope sourceScope = analyzeFrom(node, scope);
+
+            node.getWhere().ifPresent(where -> analyzeWhere(node, sourceScope, where));
+
+            List<Expression> outputExpressions = analyzeSelect(node, sourceScope);
+            List<List<Expression>> groupByExpressions = analyzeGroupBy(node, sourceScope, outputExpressions);
+            analyzeHaving(node, sourceScope);
+
+            Scope outputScope = computeAndAssignOutputScope(node, sourceScope);
+
+            List<Expression> orderByExpressions = emptyList();
+            Optional<Scope> orderByScope = Optional.empty();
+            if (node.getOrderBy().isPresent()) {
+                orderByScope = Optional.of(computeAndAssignOrderByScope(node.getOrderBy().get(), sourceScope, outputScope));
+                orderByExpressions = analyzeOrderBy(node, orderByScope.get(), outputExpressions);
+            }
+            else {
+                analysis.setOrderByExpressions(node, emptyList());
+            }
+
+            List<Expression> sourceExpressions = new ArrayList<>();
+            sourceExpressions.addAll(outputExpressions);
+            node.getHaving().ifPresent(sourceExpressions::add);
+
+            List<FunctionCall> aggregations = analyzeAggregations(node, sourceScope, orderByScope, groupByExpressions, analysis.getColumnReferences(), sourceExpressions, orderByExpressions);
+            analyzeWindowFunctions(node, outputExpressions, orderByExpressions);
+
+            if (!groupByExpressions.isEmpty() && node.getOrderBy().isPresent()) {
+                computeAndAssignOrderByScopeWithAggregation(node.getOrderBy().get(), outputScope, aggregations, groupByExpressions);
+            }
+
+            return outputScope;
+        }
+
+        private Scope legacyVisitQuerySpecification(QuerySpecification node, Optional<Scope> scope)
```

Comment:
legacy ORDER BY doesn't allow complex references to output scope fields from ORDER BY expressions. It's required by FB queries